design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,497 @@
+```
+
+When users apply this yaml file into the system.
+
+```shell script
+$ kubectl apply -f component.yaml
+```
+
+A `ControllerRevision` object will be created.
+
+```shell script
+$ kubectl get controllerrevisions
+NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
+default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
+
+$ kubectl get controllerrevisions frontend-c8bb659c5 -o yaml
+apiVersion: apps/v1
+kind: ControllerRevision
+metadata:
+  name: frontend-c8bb659c5 # you could name this anything you wanted, but just appending the semver would be good practice
+revision: 1
+data:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash lscpu"
+```
+
+And it will be recorded in the status of `component.yaml`.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: frontend
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash lscpu"
+status:
+  latestRevision: frontend-c8bb659c5
+```
+
+
+If you make a change to `component.yaml`:
+
+```shell script
+           cmd:
+-            - "bash lscpu"
++            - "bash top"
+```
+
+A new `ControllerRevision` will be automatically generated:
+
+```shell script
+$ kubectl get ControllerRevisions
+NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
+default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
+default       frontend-a75588698    core.oam.dev/component   2          2d14h
+
+$ kubectl get ControllerRevisions frontend-a75588698 -o yaml
+apiVersion: apps/v1
+kind: ControllerRevision
+metadata:
+  name: frontend-a75588698 # you could name this anything you wanted, but just appending the semver would be good practice
+revision: 2
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash top"
+```
+
+Now with `ControllerRevision` in place, let's talk about how to use it in Trait and ApplicationConfiguration.
+
+### ControllerRevision could be referenced by Trait 
+
+Some traits can control version of component which means they can specify a component revision and won't be affected
+by the change of component.
+
+For these traits, they have to declare `createRevision` field to be `true` in TraitDefinition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: fancytraits.core.oam.dev
+spec:
+  createRevision: true
```

Comment:
`createRevision` is a procedural name. Maybe rename to `revisionEnabled`

Reply:
i'm fine